examples/recurrent_attention_machine_translation.py

```diff
@@ -0,0 +1,962 @@
+    using attention.
+
+Main steps:
+- Input and target sentences are tokenized (word level).
+- Sentences are mapped to word embedding sequences.
+- An Encoder encodes the input sequence using a Bidirectional GRU
+    into a _sequence_ of encodings - with same length as the input
+    sequence.
+- A Decoder is constructed using a (unidirectional) GRU with attention:
+    At each time step:
+    - It is fed the previous target word embedding and previous GRU state.
+    - Based on the previous state and the input encoding sequence, it
+        computes a weight for each time step (i.e. word index) of the
+        input encoding sequence, using a single hidden layer MLP. The
+        attention encoding is taken as the weighted sum over the input
+        encoding sequence using these weights. (NOTE: this way the
+        Decoder has access to all time steps of the input for every
+        generated output word.)
+    - The attention encoding is concatenated to the previous target word
+        embedding and fed as input to the regular GRUCell transform.
+    - The output (updated state) of the GRUCell is fed to a single
+        hidden (maxout) layer MLP which outputs the probabilities of the
+        next target word.
+- In inference mode (greedy approach),
+    - Encode the input sequence
+    - Start with a target sequence of size 1
+        (just the start-of-sequence character)
+    - Feed the encoded input and 1-char target sequence
+        to the decoder and predict probabilities for the next word
+    - Select the next word using argmax of probabilities
+    - Append the predicted word to the target sequence
+    - Repeat until we generate the end-of-sequence character or we
+        hit the word limit.
+
+# Data
+We use the machine translation dataset described in [2].
+To download the data run:
+    mkdir -p data/wmt16_mmt
+    wget http://www.quest.dcs.shef.ac.uk/wmt16_files_mmt/training.tar.gz
+    wget http://www.quest.dcs.shef.ac.uk/wmt16_files_mmt/validation.tar.gz
+    wget http://www.quest.dcs.shef.ac.uk/wmt16_files_mmt/mmt16_task1_test.tar.gz
+    tar -xf training.tar.gz -C data/wmt16_mmt && rm training.tar.gz
+    tar -xf validation.tar.gz -C data/wmt16_mmt && rm validation.tar.gz
+    tar -xf mmt16_task1_test.tar.gz -C data/wmt16_mmt && rm mmt16_task1_test.tar.gz
+
+# References
+[1] Neural Machine Translation by Jointly Learning to Align and Translate
+    https://arxiv.org/abs/1409.0473
+[2] Multi30K: Multilingual English-German Image Descriptions
+    https://arxiv.org/abs/1605.00459
+    (http://www.statmt.org/wmt16/multimodal-task.html)
+
+# Differences between this implementation and [1]
+- NOTE that a different dataset (wmt14) is used in [1], which is _orders of
+    magnitude_ larger than the dataset used here (348M vs 0.35M words). The model
+    in [1] was trained for 252 hours (!) on a Tesla Quadro K6000, whereas for the
+    data in this example the model starts to overfit after < 1 hour (15 epochs)
```

Comment:
is `<` a typo?

Reply:
no, I meant "less than", but better type it out.